lib/msal-browser/src/app/PublicClientApplication.ts

```diff
@@ -3,281 +3,150 @@
-     * If your application supports Accounts in any organizational directory and personal Microsoft accounts, replace "Enter_the_Tenant_Info_Here" value with common.
-     * To restrict support to Personal Microsoft accounts only, replace "Enter_the_Tenant_Info_Here" value with consumers.
-     *
-     * In Azure B2C, authority is of the form https://{instance}/tfp/{tenant}/{policyName}/
-     * Full B2C functionality will be available in this library in future versions.
-     *
-     * @param configuration object for the MSAL PublicClientApplication instance
-     */
-    constructor(configuration: Configuration) {
-        super(configuration);
+    public static async createPublicClientApplication(configuration: Configuration): Promise<IPublicClientApplication> {
 
-        this.activeSilentTokenRequests = new Map();
-        // Register listener functions
-        this.trackPageVisibility = this.trackPageVisibility.bind(this);
+        const factory = new ControllerFactory(configuration);
+        const controller = await factory.createController();
+        const pca = new PublicClientApplication(configuration, controller);
+        
+        return pca;
     }
 
-    /**
-     * Use when initiating the login process by redirecting the user's browser to the authorization endpoint. This function redirects the page, so
-     * any code that follows this function will not execute.
-     *
-     * IMPORTANT: It is NOT recommended to have code that is dependent on the resolution of the Promise. This function will navigate away from the current
-     * browser window. It currently returns a Promise in order to reflect the asynchronous nature of the code running in this function.
-     *
-     * @param request
-     */
-    async loginRedirect(request?: RedirectRequest): Promise<void> {
-        const correlationId: string = this.getRequestCorrelationId(request);
-        this.logger.verbose("loginRedirect called", correlationId);
-        return this.acquireTokenRedirect({
-            correlationId,
-            ...(request || DEFAULT_REQUEST)
-        });
+    public constructor(configuration: Configuration, controller?: IController){
+        if(controller){
+            this.controller = controller;
+        }else{
+        
+            /*
+             * If loaded in an environment where window is not available,
+             * set internal flag to false so that further requests fail.
+             * This is to support server-side rendering environments.
+             */
+            const isBrowserEnvironment = typeof window !== "undefined";
+
+            const config = buildConfiguration(configuration, isBrowserEnvironment);
+            const logger = new Logger(config.system.loggerOptions, name, version);
+            const standardOperatingContext = new StandardOperatingContext(logger, config);
+            this.controller = new StandardController(standardOperatingContext, standardOperatingContext.getConfig());
```

Comment:
Can the controller get the config from the operating context itself rather than passing it in?

Reply:
+1 I think passing single `standardOperatingContext` param should be enough. `standardOperatingContext.getConfig()` does not do any manipulations with the config. Thus, we don't need an additional param for DI.

Reply:
Sure.  Don't recall why i had this way to start....

Reply:
Ok.  This is necessary when initializing the controllers from the controller factory... not needed when creating directly in PCA.

Reply:
Not sure I understand this. Are you saying the config in the OperatingContext may be different from the config passed into the Controller constructor? Is there a reason for that?

Reply:
The config only contains the config for the logger not the logger itself. The name/version is passed into the logger constructor.

Reply:
Yes I know..... this code is only here because I was told that we were not going to have any breaking changes in v3... it's going to be removed in a future PR (I think from Konstantin) who's adding an aysnc factory method and making the constructor private.  In the controllerfactory.... current OperatingContexts will need a valid logger reference even if everything else in the config is broken.  I'm actually not sure what happens when you provide an invalid configuration.... but hopefully we log when that happens....

Reply:
At the moment, initialize of the operating contexts are not doing anything.  That will change in the future.

Reply:
Sure I think we're in agreement but I don't feel like my question is being answered...I'm confused why we're passing in `operatingContext` and then also passing in a config which is retrieved by calling `operatingContext.getConfig()`. Why can the controller not call `operatingContext.getConfig()` internally? What am I missing?

Reply:
I"m pretty sure to minimize changes to the existing implementation of Client/PublicClientApplication which moved to StandardController.... I'm not msal.js expert in terms of the implementation there.  If you think it's safe to do so, I can make the larger change.  Was trying to minimize the impact on the existing implementation wherever possible.

Reply:
I see...`PublicClientApplication` is still the public API surface. The controllers are, not only new, but implementation details, so we have the freedom to design those however we want as long as `PublicClientApplication` doesn't change.

Reply:
Ok.  I'll make this change then.  I think i'm familiar enough with the code to feel comfortable doing that.

Reply:
I made some larger changes here since I was going back in anyway... let me know what you think.